subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt

```diff
@@ -240,11 +240,13 @@ fi
 # Add default JVM options here. You can also use JAVA_OPTS and ${optsEnvironmentVar} to pass JVM options to this script.
 DEFAULT_JVM_OPTS=${defaultJvmOpts}
 
-# Collect all arguments for the java command;
-#   * \$DEFAULT_JVM_OPTS, \$JAVA_OPTS, and \$${optsEnvironmentVar} can contain fragments of
-#     shell script including quotes and variable substitutions, so put them in
-#     double quotes to make sure that they get re-expanded; and
-#   * put everything else in single quotes, so that it's not re-expanded.
+
+# Collect all arguments for the java command:
+#   * $DEFAULT_JVM_OPTS, $JAVA_OPTS, and $optsEnvironmentVar can contain fragments of 
+#     shell script, and the Gradle Wrapper actively escapes special characters to ensure proper handling;
+#   * Note: The shellness will NOT be re-expanded
+#   * For example: A user cannot expect ${Hostname} to be expanded, as it is an environment variable, And will be
+#     treated as '${Hostname}' itself by the wrapper.
```

Comment:
Thanks for bringing this up @Gauravpadam.  It's intentional that this doesn't work.  I suspect we might have changed the behavior after this comment was written and missed updating it.  I think there are few minor changes we should make, though.

1. First, the `\` before `$` was intentional as this file is a template rather than a script itself.  Removing the `\` will cause Groovy to evaluate the string as a template variable.  However, instead of adding the `\` back, let's just remove the `$` altogether.  It doesn't add anything to the comment and opens us up to accidentally removing the `\` again in the future.
2. I find the wording still is not 100% clear what is allowed in these variables.  Let's just say directly that DEFAULT_JVM_OPTS, JAVA_OPTS, JAVA_OPTS, and optsEnvironmentVar are not allowed to contain shell fragments and that any embedded shellness will be escaped.  Your example is good, I just want to tighten up the preceding comments to be more direct and unambiguous.
3. I believe this script is used for things other than the wrapper (like the scripts generated by the Application plugin).  Instead of `treated as '${Hostname}' itself by the wrapper` let's change it to `treated as '${Hostname}' itself on the command line`

Reply:
Thanks of the insights @ghale, I've made the required changes in the latest commit